src/StrawberryfieldFilePersisterService.php

```diff
@@ -191,39 +193,51 @@ public function getDestinationUri(
     // If no destination scheme was setup on our global config use the original file scheme.
     $desired_scheme = !empty($this->destinationScheme) ? $this->destinationScheme : $file_parts['destination_scheme'];
 
-    $destination_filename = $file_parts['destination_filetype'] . '-' . $file_parts['destination_filename'] . '.' . $file_parts['destination_extension'];
+    // Mime type becomes suffix. Performant for filtering in S3.
+    $destination_basename = $file_parts['destination_filetype'] . '-' . $file_parts['destination_filename'];
 
-    // Sanitize filename.
-    // @see http://stackoverflow.com/questions/2021624/string-sanitizer-for-filename
-    // Do this always, why would i not want this?
+    // Sanitize the whole thing.
+    $destination_basename = $this->sanitizeFileName($destination_basename);
 
-    $destination_basename = substr(
-      pathinfo($destination_filename, PATHINFO_BASENAME),
-      0,
-      -strlen(".{$file_parts['destination_extension']}")
-    );
-    $destination_basename = mb_strtolower($destination_basename);
-    $destination_basename = $this->transliteration->transliterate(
-      $destination_basename,
-      $this->languageManager->getCurrentLanguage()->getId(),
-      '-'
-    );
-    $destination_basename = preg_replace(
-      '([^\w\s\d\-_~,;:\[\]\(\)\].]|[\.]{2,})',
-      '',
-      $destination_basename
-    );
-    $destination_basename = preg_replace('/\s+/', '-', $destination_basename);
-    $destination_basename = trim($destination_basename, '-');
-    // If the basename if empty use the element's key.
+    // Edge case, should rarely happen.
     if (empty($destination_basename)) {
-      $destination_basename = $file_parts['destination_filetype'] . '-' . $file->uuid(
-        );
+      $destination_basename = $file_parts['destination_filetype'] . '-unnamed';
```

Comment:
@DiegoPino If empty $destination_basename it means $file_parts['destination_filetype'] id empty or empty after sanitize so I suggest :
`$destination_basename = 'nomimetype' . '-unnamed';`

Reply:
HI, yep, you are right. My logic there was so wrong, mind still not totally active. Fixing now.
I will go for 'application' instead of 'nomimetype' just to conform with how we calculate that part when not empty which is first part of the mime type and since unknown mime-type is always application/octet-stream we end having application. That if is really an overkill, since it assuming that Drupal would end having a null in the mime-type and the filename would have only invalid characters, which would all happen at the same time if something goes terrible wrong. But i like to be strict on this since we are dealing with valuable repository items.

Reply:
@giancarlobi this should be ready. If you have some time tomorrow to give this a look it would be great! Thanks so much for your feedback =)

Reply:
It looks good, great.